src/core/server/http/http_server.ts

```diff
@@ -72,6 +77,28 @@ export class HttpServer {
     this.registeredRouters.add(router);
   }
 
+  // passing hapi Request works for BWC. can be deleted once we remove legacy server.
+  private getBasePathFor(config: HttpConfig, request: KibanaRequest | Request) {
+    const incomingMessage =
+      request instanceof KibanaRequest ? request.unstable_getIncomingMessage() : request.raw.req;
+
+    const requestScopePath = this.basePathCache.get(incomingMessage) || '';
+    const serverBasePath = config.basePath || '';
+    return `${serverBasePath}${requestScopePath}`;
+  }
+
+  // should work only for KibanaRequest as soon as spaces migrate to NP
+  private setBasePathFor(request: KibanaRequest | Request, basePath: string) {
```

Comment:
We store basePath in WeakMap, where `request.raw.req: IncomingMessage` is used as a key. Hapi re-uses `request.raw.req` when we proxy request to LP, so it still can be used as a key to [retrieve basePath from NP api](https://github.com/elastic/kibana/blob/d7bbc912e375fda24b9e266adb159dadc7a97e40/src/legacy/server/http/setup_base_path_provider.js)

Reply:
Can I ask why we opt for a WeakMap instead of a Map for this cache?

Reply:
If we use ordinary Map it will lead to memory leaks, because Map keeps a reference to incoming request and it cannot be removed by GC. Using `WeakMap` frees us from the obligation to control the allocated memory manually.